flash/core/integrations/vissl/base_ssl_model.py

```diff
@@ -0,0 +1,584 @@
+                    head_type.append(head_param[idx][0])
+                head_name = "->".join(head_type)
+                head = nn.Sequential(*head_modules)
+            else:
+                # head is a single module
+                head_name = head_param[0]
+                with set_torch_seed(self.model_config.MODEL_INIT_SEED):
+                    head = self._build_head_module(head_param)
+            self.heads.append(head)
+            self.head_names.append(head_name)
+
+    def _setup_multi_input_head_mapping(self):
+        """
+        Used for PIRL style training where the model operates
+        on image and the patches.
+
+        Assumptions:
+        - This assumes that the same trunk is used to extract features
+          for the different types of inputs.
+        - One head only operates on one kind of input, Every individual
+          head can contain several layers. See _get_heads() function for examples.
+
+        * Specify Input -> Trunk Features mapping
+           Like in the single input case, the heads can operate on features
+           from different layers. In this case, we specify MODEL.MULTI_INPUT_HEAD_MAPPING
+           to be a list like:
+           [
+               ["input_key", [list of features heads is applied on]]
+           ]
+           For example: for a model that applies two heads on images
+                        and one head on patches
+                        [
+                            ["images", ["res5", "res4"]],
+                            ["patches", ["res3"]],
+                        ]
+        """
+        if len(self.model_config.MULTI_INPUT_HEAD_MAPPING) == 0:
+            return
+
+        assert len(self.model_config.MULTI_INPUT_HEAD_MAPPING) == len(
+            self.heads
+        ), "MULTI_INPUT_HEAD_MAPPING must be a list of length == #heads"
+
+        # create many-to-one mapping from input_key to head
+        self._input_to_head_map = {}
+        for idx in range(len(self.model_config.MULTI_INPUT_HEAD_MAPPING)):
+            key = self.model_config.MULTI_INPUT_HEAD_MAPPING[idx][0]
+            if key not in self._input_to_head_map:
+                self._input_to_head_map[key] = []
+            self._input_to_head_map[key].append(self.heads[idx])
+
+        # create many-to-one mapping from input key to eval features
+        self._input_to_eval_features_map = {}
+        for input_idx in range(len(self.model_config.MULTI_INPUT_HEAD_MAPPING)):
+            key = self.model_config.MULTI_INPUT_HEAD_MAPPING[input_idx][0]
+            eval_layer_names = self.model_config.MULTI_INPUT_HEAD_MAPPING[input_idx][1]
+            if key in self._input_to_eval_features_map:
+                raise ValueError(
+                    f"duplicate key {key} \
+                    specified for MODEL.MULTI_INPUT_HEAD_MAPPING."
+                )
+            self._input_to_eval_features_map[key] = eval_layer_names
+
+    def get_classy_state(self, deep_copy=False):
```

Comment:
I don't think the model should contain this. Lightning should be responsible to handle the checkpoiting.